### PR TITLE
 Update to AnVIL schema v4 (#4617)

### LINKED ISSUE #4617: Update to AnVIL schema v4
https://github.com/broadinstitute/anvil_tdr_ingest/pull/2/files

- [x] Switch the v3 snapshots in the corpus to their updated v4 counterparts
- [x] Absorb renaming of fields
- [x] Absorb renaming of tables
- [ ] Ensure that we index all tables
- [ ] Ensure that we follow all foreign keys in the schema
- [x] Ensure that we index all columns in the schema

I don't think we currently follow the FK `from_biosample.diagnosis_id_to_diagnosis.diagnosis_id` that's being removed in v4. So nothing to do on that front. But see #4618.

Absorbing the renamed tables doesn't affect our internal naming of entity types i.e. `biosample` stays `biosample` even though the table is called `anvil_biosample`.

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #4617 

### Author

- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue        <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via Zenhub 
- [x] PR description links to connected issues
- [x] Added `partial` label to PR                                   <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain               <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR                             <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues      <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased PR branch on `develop`, squashed old fixups

### Peer reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists. Update the `N reviews` label. 

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations         <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or all connected issues are labeled `no demo`</sub>
- [x] Rebased and squashed PR branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label    <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment                          <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build log for anomalies in `sandbox` deployment      <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev`                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build log for anomalies in `anvilbox` deployment     <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `anvilbox`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `anvilbox`                                 <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox`                            <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain                                        <sub>or this PR is not labeled `base`</sub>
- [x] Pushed merge commit to GitLab `dev`                           <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev`                      <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build log for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build log for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is* 
pushed determines this checklist item.

### Operator (reindex) 

- [x] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Deleted unreferenced indices in `anvildev`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev`                                 <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev`            <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev`       <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment                       <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment                  <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Review by achave11-ucsc — APPROVED** [review]
LGTM 👌🏽

File: scripts/can_bundle.py [hunk 1]
```diff
@@ -59,8 +67,12 @@ def main(argv):
     parser.add_argument('--output-dir', '-O',
                         default=os.path.join(config.project_root, 'test', 'indexer', 'data'),
                         help='The path to the output directory (default: %(default)s).')
+    parser.add_argument('--redaction-key', '-K',
+                        help='Provide a key to redact personally senstive information from the output files')
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py]:
```suggestion
                        help='Provide a key to redact confidential or sensitive information from the output files')
```

File: scripts/can_bundle.py [hunk 2]
```diff
@@ -100,6 +112,58 @@ def save_bundle(bundle: Bundle, output_dir: str) -> None:
...
+        if entity_type in redacted_entity_types:
+            redact_json(entity, key)
+
+
+def redact_json(o: AnyMutableJSON, key: bytes) -> AnyMutableJSON:
+    """
+    >>> key = b'bananas'
+    >>> redact_json('sensitive', key)
+    'redacted-AVm0tjOw'
+    >>> redact_json('sensitive', key + b'plit')
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py:134]: Empty line between test cases.

File: scripts/can_bundle.py [hunk 3]
```diff
@@ -100,6 +112,58 @@ def save_bundle(bundle: Bundle, output_dir: str) -> None:
...
+        try:
+            uuid.UUID(o)
+        except ValueError:
+            o = base64.b64encode(hashlib.sha1(key + o.encode()).digest()[:6])
+            o = (b'redacted-' + o).decode()
+    elif isinstance(o, (int, float)):
+        o = struct.unpack('>Q', hashlib.sha1(key + str(o).encode()).digest()[:8])[0]
+        o = (o & 0xFFFFFFFFFFFF) + 42000000000000000
+    elif isinstance(o, list):
+        o[:] = sorted(redact_json(item, key) for item in o)
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py]: A comment should explain the need for sorting.

File: scripts/can_bundle.py [hunk 4]
```diff
@@ -100,6 +112,58 @@ def save_bundle(bundle: Bundle, output_dir: str) -> None:
...
+
+def redact_bundle(bundle: Bundle, key: bytes) -> None:
+    for name, entity in bundle.metadata_files.items():
+        entity: MutableJSON
+        entity_type = name.split('_')[0]
+        if entity_type in redacted_entity_types:
+            redact_json(entity, key)
+
+
+def redact_json(o: AnyMutableJSON, key: bytes) -> AnyMutableJSON:
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py]: Either return or mutate in place, never both. If a mix is needed for recursion, a wrapper should hide that. I personally find pure functions easier to write and reason about. I only resort non-pure for performance reasons, or when that is easier to understand, for example, when a surgical change is made. This is more of a transformation so I would write a pure function.

FYI  @achave11

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I am confused about the files names and provenance of the cans. 

This comment sounds wrong to me: https://github.com/DataBiosphere/azul/blob/64ae91b57af6a0f6fd2001fc2bf2a73aa4e53618/scripts/recan_bundle_tdr.py#L358

The deletion of `.manifest.json` and `.metadata.json` should be an early, separate commit if those files are truly redundant.

The `.result` and `.results` in the file names are confusing, very similar but different semantics. I am also confused as to what purpose the `.tables` file has.

And there's a smell with https://github.com/DataBiosphere/azul/blob/64ae91b57af6a0f6fd2001fc2bf2a73aa4e53618/scripts/recan_bundle_tdr.py#L77

I think the code to create it is longer than a literal of the result, even though there is no advantage to making it dynamic since the references are very static.

Please request a PL slot for this.

File: scripts/can_bundle.py [hunk 5]
```diff
@@ -100,6 +112,64 @@ def save_bundle(bundle: Bundle, output_dir: str) -> None:
...
+def redact_bundle(bundle: Bundle, key: bytes) -> None:
+    for name in bundle.metadata_files.keys():
+        entity_type = name.split('_')[0]
+        if entity_type in redacted_entity_types:
+            bundle.metadata_files[name] = redact_json(bundle.metadata_files[name], key)
+
+
+def redact_json(o: AnyJSON, key: bytes) -> AnyMutableJSON:
+    """
+    >>> key = b'bananas'
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py:130]:
```suggestion
    >>> key = b'bananas'
    
```

File: scripts/can_bundle.py (hunk 1 above)
Comment (hannes-ucsc) [reply scripts/can_bundle.py]: Please address https://github.com/DataBiosphere/azul/pull/4741/files#r1038456379

File: scripts/can_bundle.py (hunk 5 above)
Comment (hannes-ucsc) [reply scripts/can_bundle.py:130]: This script should be included in `test_doctests.py`

File: scripts/can_bundle.py [hunk 6]
```diff
@@ -100,6 +112,64 @@ def save_bundle(bundle: Bundle, output_dir: str) -> None:
...
+    'redacted-AVm0tjOw'
+
+    >>> redact_json('sensitive', key + b'plit')
+    'redacted-+L3zz1rW'
+
+    >>> redact_json(123, key)
+    42027752232213208
+
+    >>> redact_json(['sensitive', 123], key)
+    ['redacted-AVm0tjOw', 42027752232213208]
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py]: These can't be passing since Python can't sort heterogeneous iterables.

File: scripts/can_bundle.py [hunk 7]
```diff
@@ -100,6 +112,64 @@ def save_bundle(bundle: Bundle, output_dir: str) -> None:
...
+
+    >>> redact_json({'foo': ['sensitive', {'bar': 123}]}, key)
+    {'foo': ['redacted-AVm0tjOw', {'bar': 42027752232213208}]}
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py]: These can't be passing since Python can't sort heterogeneous iterables.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: scripts/can_bundle.py [hunk 8]
```diff
@@ -59,8 +67,12 @@ def main(argv):
     parser.add_argument('--output-dir', '-O',
                         default=os.path.join(config.project_root, 'test', 'indexer', 'data'),
                         help='The path to the output directory (default: %(default)s).')
+    parser.add_argument('--redaction-key', '-K',
+                        help='Provide a key to redact confidential or senstive information from the output files')
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py]:
```suggestion
                        help='Provide a key to redact confidential or sensitive information from the output files')
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**